expression/builtin_compare.go

```diff
@@ -1079,37 +1085,48 @@ func tryToConvertConstantInt(ctx sessionctx.Context, isUnsigned bool, con *Const
 		return con, false
 	}
 	sc := ctx.GetSessionVars().StmtCtx
-	fieldType := types.NewFieldType(mysql.TypeLonglong)
-	if isUnsigned {
-		fieldType.Flag |= mysql.UnsignedFlag
-	}
-	dt, err = dt.ConvertTo(sc, fieldType)
+
+	dt, err = dt.ConvertTo(sc, targetFieldType)
 	if err != nil {
-		return con, terror.ErrorEqual(err, types.ErrOverflow)
+		if terror.ErrorEqual(err, types.ErrOverflow) {
+			return &Constant{
+				Value:   dt,
+				RetType: targetFieldType,
+			}, true
+		}
+		return con, false
 	}
 	return &Constant{
 		Value:        dt,
-		RetType:      fieldType,
+		RetType:      targetFieldType,
 		DeferredExpr: con.DeferredExpr,
 	}, false
 }
 
-// RefineComparedConstant changes an non-integer constant argument to its ceiling or floor result by the given op.
-// isAlwaysFalse indicates whether the int column "con" is false.
-func RefineComparedConstant(ctx sessionctx.Context, isUnsigned bool, con *Constant, op opcode.Op) (_ *Constant, isAlwaysFalse bool) {
+// RefineComparedConstant changes a non-integer constant argument to its ceiling or floor result by the given op.
+// isExceptional indicates whether the 'int column [cmp] const' might be true/false.
+// If isExceptional is true, CorrectVal is returned. Or, ExecptionalVal is returned.
+// CorrectVal: The computed result. If the constant can be converted to int without exception, return the val. Else return 'con'(the input).
+// ExceptionalVal : It is used to get more information to check whether 'int column [cmp] const' is true/false
+// 					If the op == LT,LE,GT,GE and it gets an Overflow when converting, return inf/-inf.
+// 					If the op == EQ,NullEQ and the constant can never be equal to the int column, return the CorrectVal.
+func RefineComparedConstant(ctx sessionctx.Context, targetFieldType types.FieldType, con *Constant, op opcode.Op) (_ *Constant, isExceptional bool) {
 	dt, err := con.Eval(chunk.Row{})
 	if err != nil {
 		return con, false
 	}
 	sc := ctx.GetSessionVars().StmtCtx
-	intFieldType := types.NewFieldType(mysql.TypeLonglong)
-	if isUnsigned {
-		intFieldType.Flag |= mysql.UnsignedFlag
-	}
+
 	var intDatum types.Datum
-	intDatum, err = dt.ConvertTo(sc, intFieldType)
+	intDatum, err = dt.ConvertTo(sc, &targetFieldType)
 	if err != nil {
-		return con, terror.ErrorEqual(err, types.ErrOverflow)
+		if terror.ErrorEqual(err, types.ErrOverflow) {
+			return &Constant{
+				Value:   intDatum,
```

Comment:
If overflow happens, what value is the `intDatum`?
Can we guarantee that we can use `intDatum.Value.GetInt64()&1` to judge if it is INF when overflow happens?